### ISSUE TITLE
Visual Studio 2022 Support

### ISSUE THREAD

**Issue by scottkuhl** [microsoft/TemplateStudio#4353]
### Installed product versions
- Visual Studio: 2022 All Versions

### Description
Make the extension available in Visual Studio 2022.

### Steps to recreate
1. Download the extension installer.
2. Run the installer.

### Current behavior
Message that it is already installed to all applicable products appears even though it is not installed in Visual Studio 2022.
It is also not listed in Online Extensions in Manage Extensions Visual Studio 2022.

### Expected behavior
Installs and works in Visual Studio 2022.

_Note: Sorry for the all the extra details.  I probably could have stopped with the title on this one._

**Comment by scottkuhl** [comment]
[Writing Visual Studio Extensions with Mads - Splitting extensions for Visual Studio 2022](https://www.youtube.com/watch?v=-PKIPTW6km0)

**Comment by sibille** [comment]
Thanks for opening this issue. We've looked into support for VS 2022 but need a few architecture changes before being able to support VS2019 and VS 2022 side by side.

**Comment by itsbeer** [comment]
Any updates on this? 

Its my understanding that there are libraries that have x64 architecture issues that are being waited on before an updated extension can be released. Is that still the situation? If so, what exactly are the libraries at issue? 

With that knowledge, maybe the community at large can help speed things along. This is hands down one of my favorite and most useful extensions. It would be a shame for the VS2022 user base not to benefit from it.

**Comment by mikebattista** [comment]
I'm digging into this at the moment. Whatever remaining blocking issues I find, I will list them here.

**Comment by mrlacey** [comment]
> I'm digging into this at the moment. Whatever remaining blocking issues I find, I will list them here.

Here's a new issue to add to the list: https://developercommunity.visualstudio.com/t/creating-project-dialog-is-always-on-top-in-vs2022/1608136

**Comment by mikebattista** [comment]
Thanks. I did get things running in the latest 2022 preview and noticed this bug as well. I'm working in [my fork](https://github.com/mikebattista/WindowsTemplateStudio) if you want to take a look. It needs more validation.

The 2019 and 2022 extensions will need different IDs and Marketplace listings for now per https://docs.microsoft.com/en-us/visualstudio/extensibility/migration/update-visual-studio-extension?view=vs-2022#add-a-visual-studio-2022-target and https://docs.microsoft.com/en-us/visualstudio/extensibility/migration/update-visual-studio-extension?view=vs-2022#visual-studio-marketplace. In the future the IDs can be the same and then listed under a merged listing.

I looked at https://docs.microsoft.com/en-us/visualstudio/extensibility/migration/update-visual-studio-extension?view=vs-2022#use-shared-projects-for-multi-targeting initially but ran into some complexity and limitations in the tooling from my own experience and from others that have updated their extensions. The "architecture changes" mentioned earlier on this thread were related to refactoring for the shared project approach.

An alternative approach that was recommended is to just use separate branches for the 2022 and 2019 extensions. Work would happen in the 2022 branch and then be merged into the 2019 branch.

**Comment by mrlacey** [comment]
For reference. there are also requests for this in marketplace reviews: https://marketplace.visualstudio.com/items?itemName=WASTeamAccount.WindowsTemplateStudio&ssr=false#review-details

**Comment by itsbeer** [comment]
Any one have any updates on this? Appreciate everyone's work on it. @mrlacey @sibille and others.

**Comment by mrlacey** [comment]
> Any one have any updates on this? Appreciate everyone's work on it. @mrlacey @sibille and others.

As per https://github.com/microsoft/WindowsTemplateStudio/issues/4383#issuecomment-1008433018 , I hope to have more news soon. Meanwhile, have a look at https://github.com/mrlacey/TemplateStudio

**Comment by mikebattista** [comment]
We've been working on a roadmap for the project which we'll publish soon for review.

Regarding supporting VS 2022, there are a few strategies regarding how to support both VS 2019 and VS 2022 that we are evaluating. Once we close on that and some other parts of the roadmap, we'll be able to move forward with this.

>  Meanwhile, have a look at https://github.com/mrlacey/TemplateStudio

As mentioned above, @mrlacey made some tactical changes in his fork to unblock in the meantime.

**Comment by itsbeer** [comment]
Thanks for the update @mrlacey and @mikebattista really appreciate it! I am going to play around with Matt's project for a bit until movement is made here.

**Comment by mikebattista** [comment]
We just published our updated [roadmap](https://github.com/microsoft/TemplateStudio/blob/dev/docs/roadmap.md) for review. Let us know here if you have any feedback.

**Comment by itsbeer** [comment]
@mikebattista Thats awesome good news! Any timeline for the implementation?

**Comment by mikebattista** [comment]
@mrlacey has been refactoring the project to produce separate extensions in his fork. We'll be looking at that over the next few weeks to merge the code into this repo. That refactor is the biggest hurdle right now so will take some time. @mrlacey can provide more details on progress there.

Once that's completed, we'd want to publish English-only versions of the 2022 extensions as soon as possible with the existing templates that are there to unblock 2022 development. Then we would start updating to .NET 6.0, Windows App SDK 1.0, adding features, adding back localization, etc. in future updates. We can probably work on the template updates in parallel in a separate branch while we bring up the baseline 2022 builds.

If I had to estimate a timeline for that initial VS 2022 release of all the extensions, I'd guess 4-8 weeks but it really depends on how the refactor goes. Once that initial release is out, though, we should be able to turn around updates to the templates much faster.

**Comment by itsbeer** [comment]
That is FANTASTIC news @mikebattista! I like the direction things are going in with the separate packages / projects and with the name change to better support a broader set of platforms and use cases going forward. Appreciate the work and effort from everyone involved! Can't wait for the bits to a release candidate to hit.

**Comment by mikebattista** [comment]
WinUI and WPF templates have been updated as well. Just waiting to merge that in with Matt's refactor work.

https://github.com/microsoft/TemplateStudio/issues/4397
https://github.com/microsoft/TemplateStudio/issues/4398
https://github.com/microsoft/TemplateStudio/issues/4399
https://github.com/microsoft/TemplateStudio/issues/4401

**Comment by mrlacey** [comment]
> Status update: I've found a fix for my last remaining blocking issue. I expect to be able to start making this public next week. 😀

Status update: Getting testing working is more complicated and therefore taking longer than expected.

**Comment by mrlacey** [comment]
> WinUI and WPF templates have been updated as well. Just waiting to merge that in with Matt's refactor work.
> 
> #4397 #4398 #4399 #4401

Note. This will not be a straight merge due to significant internal refactoring and how templates are included in separate extension projects.

**Comment by mrlacey** [comment]
Quick question regarding casing. Which is preferred/recommended for distinguishing between the C# & C++ versions of the WinUI templates (in names)?

1. WinuiCs & WinuiCpp
2. WinUiCs & WinUiCpp
3. WinUiCS & WinUiCPP
4. WinUICs & WinUICpp
5. WinUICS & WinUICPP
6. some other suggestion...

**Comment by mikebattista** [comment]
Where will these names show up?

**Comment by mrlacey** [comment]
> Where will these names show up?

in various/numerous places in the code of the project/wizard: project names, namespaces, internal classes, tests, etc.
Not in generated code.

I'm just looking to be consistent and meet expectations. It'll be a pain to change them (again)

**Comment by mikebattista** [comment]
WinUI for sure which leaves 4 and 5. I prefer 4 over 5 as a starting point. Will want to see it in context though.

**Comment by mrlacey** [comment]
I've gone for Option 4.
It means the projects in the solution are named like this:
![image](https://user-images.githubusercontent.com/189547/156408076-84a953b1-0318-4563-a965-df0e0906e7b4.png)

**Comment by mikebattista** [comment]
Looks good. Thanks.

**Comment by mcNets** [comment]
Hi.

I know Matt's templates are available in preview, but is there any ETA for the official release?

**Comment by mrlacey** [comment]
> I know Matt's templates are available in preview, but is there any ETA for the official release?

My templates will become the official ones. I'm just removing the bits I added just for me, putting back the localization, and making sure all tests pass and everything works ok.
Answer: As soon as I can. I really hope to be done by the end of this month.

There are two things you can do to speed things along:
1. Test the previews. 
2. Give feedback on the WinUI templates. Personally, I think they need a [LOT](https://github.com/mrlacey/TemplateStudio/issues/6) of changes.


[third thing: 💵](https://github.com/sponsors/mrlacey)

**Comment by mcNets** [comment]
Thanks @mrlacey.

Just a question, I'm trying to replace the standard title bar. Do you have any docs about `ExtendViewIntoTitleBar`?

**Comment by Diegorro98** [comment]
@mcNets is this useful for you? [Title bar customization (Microsoft Docs)](https://docs.microsoft.com/en-us/windows/apps/develop/title-bar?tabs=winui2#full-customization)

**Comment by mcNets** [comment]
> @mcNets is this useful for you? [Title bar customization (Microsoft Docs)](https://docs.microsoft.com/en-us/windows/apps/develop/title-bar?tabs=winui2#full-customization)

Well, what I tried is:

1-I've added a new window to the project with some content to replace the title bar. (I need to call Window.SetTitleBar())

```
<Grid>
        <!--TitleBar-->
        <Border x:Name="AppTitleBar"
                IsHitTestVisible="True"
                VerticalAlignment="Top"
                Background="Transparent"
                Height="40"
                Canvas.ZIndex="1"
                Margin="78,8,0,0">
            <StackPanel Orientation="Horizontal">
                <TextBlock x:Name="AppTitle"
                           Text="{x:Bind GetAppTitleFromSystem()}"
                           VerticalAlignment="Center"
                           HorizontalAlignment="Center"
                           Margin="12,0,0,0"
                           Style="{StaticResource CaptionTextBlockStyle}" />
            </StackPanel>
        </Border>
        <Frame x:Name="MainContent"
            x:FieldModifier="public" />
    </Grid>
```
 2-I've changed the static MainWindow by my own:
 
 `public static Window MainWindow { get; set; } = new AppWindow() { Title = "AppDisplayName".GetLocalized() };` 
 
 But I'm having problems trying to activate MainPage inside my own window:
 
```
 AppWindow wnd = (AppWindow)App.MainWindow;
_shell = new Frame();
wnd.MainContent.SourcePageType = typeof(ShellPage);

//if (App.MainWindow.Content == null)
//{
//    _shell = Ioc.Default.GetService<ShellPage>();
//    App.MainWindow.Content = _shell ?? new Frame();
//}

```

**Comment by mikebattista** [comment]
This repo is producing 2022 extensions now so closing out.

Work remaining before we ship is tracked at https://github.com/microsoft/TemplateStudio/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Visual+Studio+2022%22 which mostly involves merging WPF template updates, finishing the rebrand from Windows Template Studio to Template Studio, and updating remaining build and release pipelines to account for the new project structure.

**Comment by mikebattista** [comment]
We just shipped the new 2022 extensions to the Marketplace!

  * [Template Studio for WinUI (C#)](https://marketplace.visualstudio.com/items?itemName=TemplateStudio.TemplateStudioForWinUICs)
  * [Template Studio for WPF](https://marketplace.visualstudio.com/items?itemName=TemplateStudio.TemplateStudioForWPF)
  * [Template Studio for UWP](https://marketplace.visualstudio.com/items?itemName=TemplateStudio.TemplateStudioForUWP)

Special thanks to @mrlacey for all the contributions refactoring the extensions and updating them for VS 2022. A lot of work went into the refactor, and it provides a solid foundation we can build on going forward.

Release notes are https://github.com/microsoft/TemplateStudio/releases/tag/v5.0.

Give them a try and let us know if you have any feedback!

**Comment by itsbeer** [comment]
You folks are simply AMAZING! Thank you all for your great work and dedication! Going to uninstall Matt's testing versions and install the newly minted extensions first chance I get.

Thanks again!
